### PR TITLE
6873: UGENE crashes in "Pairwise alignment" for big alignments

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp [hunk 1]
```diff
@@ -307,7 +307,7 @@ void SmithWatermanAlgorithm::calculateMatrixForMultipleAlignmentResult() {
         }
     } while (++i <= static_cast<int>(src_n));
 
-    free(matrix);
+    delete matrix;
```
Comment (atiunov) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: delete [] matrix

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp [hunk 2]
```diff
@@ -414,7 +414,7 @@ void SmithWatermanAlgorithm::calculateMatrixForAnnotationsResult() {
     }
 #endif
 
-    free(matrix);
+    delete matrix;
```
Comment (atiunov) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: delete [] matrix

**Review by atiunov — CHANGES_REQUESTED** [review]
Just replacing malloc calls with new is not enough. There is a solution offered in the comment to the issue.

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp (hunk 2 above)
Comment (sgrekhov) [reply src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: > Just replacing malloc calls with new is not enough. There is a solution offered in the comment to the issue.

This task is called from https://github.com/ugeneunipro/ugene/blob/master/src/corelibs/U2Private/src/TaskSchedulerImpl.cpp#L1052-L1057
```C
        try {
            ti->task->run();
            assert(ti->task->getState() == Task::State_Running);
        } catch (const std::bad_alloc &) {
            onBadAlloc(ti->task);
        }
```
Calling `new` throws `bad_alloc` in case of low memory (unlike `malloc()`) which is caught here and sets an error in `onBadAlloc`

Comment (DmitriiSukhomlinov) [reply src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: I agree with @sgrekhov . There is no need to write error to the operationStatus member. If there is not enough memory to allocate, "new" throws the "bad_alloc" exception and TaskScheduler catches it.

Comment (atiunov) [reply src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: Ok, you convinced me. But fix delete calls, pointers should be released as pointers to array, not as pointers to a value.

**Review by atiunov — CHANGES_REQUESTED** [review]
Delete calls should be fixed.

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp [hunk 3]
```diff
@@ -173,7 +173,7 @@ void SmithWatermanAlgorithm::calculateMatrixForMultipleAlignmentResult() {
 
     n = pat_n * 2;
     int dirn = (4 + pat_n + 3) >> 2;
-    int *buf, *matrix = (int *)malloc(n * sizeof(int) + pat_n * 0x80 + matrixLength * dirn);
+    int *buf, *matrix = new int[n * sizeof(int) + pat_n * 0x80 + matrixLength * dirn];
```
Comment (DmitriiSukhomlinov) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: Not like this. When we use "malloc" we need to pass the exact memory value (in bytes) we need to allocate as an argument, usually it looks like:
`malloc(size_of_array * syze_of_one_element_of_array_in_bytes)`
And when we use "new" we pass just size of the array,  like:
`new[size_of_array]`
Let's figure our case out:
`malloc(n * sizeof(int) + pat_n * 0x80 + matrixLength * dirn)`
`n * sizeof(int)` - **n** is size of araray, **sizeof(int)** - sizeof is a function, which calculates size of the object (or of the type, _int_, in our case)
`pat_n * 0x80 + matrixLength * dirn` - some additional information which is stored in this array (I hate the guy who did this, he should use separate variables)
So, we need to decrease the array size, I think like this:
`new int[n + ((pat_n * 0x80 + matrixLength * dirn) / sizeof(int))];`

PS: test it!

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp (hunk 3 above)
Comment (yalgaer) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: Why can't you keep the old code and just add a check: if (matrix === nullptr) throw bad-alloc... ?

+ there is another place like this with malloc() in the file.

**Review by yalgaer — CHANGES_REQUESTED** [review]

**Comment by yalgaer** [comment]
These is another improvement can be done here that will help to make behavior more predictable and reliable.

UGENE has "AppResource" framework. A task can declare its needs (like I need 1Gb of memory). This framework is used for memory/cpu consuming tasks.

Task scheduler will not run 2 tasks in parallel if their total memory usage exceeds some MAX limit.
If resources required by a single task are above the limit - the task will not be run at all: user will see a valid error message.

So the complete solution would be adding to SW task an estimator: how much resources does it need?

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp (hunk 3 above)
Comment (yalgaer) [reply src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: new qint8[original size]  looks simpler ?

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp [hunk 4]
```diff
@@ -173,7 +173,7 @@ void SmithWatermanAlgorithm::calculateMatrixForMultipleAlignmentResult() {
...
     char *score, *score1 = (char *)(matrix + n);
     unsigned char *dir, *dir2, *dir1 = (unsigned char *)score1 + pat_n * 0x80;
     memset(matrix, 0, n * sizeof(int));
```
Comment (yalgaer) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp:183]: While this is not directly related to the fix it it looks strange to call malloc () memset(0) when you can use a single calloc() instead.
Leaving it up to you if to fix it or not.

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp (hunk 3 above)
Comment (sgrekhov) [reply src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp]: Ok. Fix with minimal changes

**Comment by sgrekhov** [comment]
> These is another improvement can be done here that will help to make behavior more predictable and reliable.
> 
> UGENE has "AppResource" framework. A task can declare its needs (like I need 1Gb of memory). This framework is used for memory/cpu consuming tasks.
> 
> Task scheduler will not run 2 tasks in parallel if their total memory usage exceeds some MAX limit.
> If resources required by a single task are above the limit - the task will not be run at all: user will see a valid error message.
> 
> So the complete solution would be adding to SW task an estimator: how much resources does it need?

After the fix the task can be run sucessfully. Seems that UGENE does several attempts to run the task and one of them is sucessful

File: src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp [hunk 5]
```diff
@@ -174,6 +174,10 @@ void SmithWatermanAlgorithm::calculateMatrixForMultipleAlignmentResult() {
     n = pat_n * 2;
     int dirn = (4 + pat_n + 3) >> 2;
     int *buf, *matrix = (int *)malloc(n * sizeof(int) + pat_n * 0x80 + matrixLength * dirn);
+    if (matrix == NULL) {
+        std::bad_alloc e;
+        throw e;
```
Comment (yalgaer) [inline src/plugins/smith_waterman/src/SmithWatermanAlgorithm.cpp:179]: not a problem: can be done in 1 line: throw std::bad_alloc();

**Review by yalgaer — APPROVED** [review]

**Review by atiunov — APPROVED** [review]
